### DISCUSSION TITLE
Debug/Test WeakReference

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#10346] [Q&A]
MvvmDialogs is nearly completed for mobile including back navigation support.

There is one remaining issue that I'm unable to find the solution. Not a deal-breaker; but not as intended either.

For mobile navigation, the idea is that you show ViewModel1 with View1, then ViewModel2 with View2, at which point View1 doesn't need to be kept in memory so it's stored in a cache with WeakReference. I'm testing to ensure that non-visible views can be unloaded by calling `GC.Collect()`, and that test fails. Views can always be reconstructed on-demand from the ViewModel.

Could someone with more experience in this help me figure out this detail? I might have something to learn here.

1. Download the [MvvmDialogs repo](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/)

2. Run the tests. [This test fails](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/blob/master/src/MvvmDialogs.Avalonia.Tests/NavigationTests.cs#L91)

3. Figure out how to make that test succeed.

What am I missing? I tried running it in Release mode, and tried a few other tricks to help it release; none of it worked. Perhaps there's a reference stuck somewhere. Or could something within Avalonia hold the View from releasing?

Most of the relevant code is in the [NavigationManager](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/blob/master/src/MvvmDialogs.Avalonia/Navigation/NavigationManager.cs) and [ViewCache](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/blob/master/src/MvvmDialogs.Avalonia/Navigation/ViewCache.cs)/[ViewCacheItem](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/blob/master/src/MvvmDialogs.Avalonia/Navigation/ViewCacheItem.cs).

**Comment by jp2masa** [comment]
Do you have dotMemory? If so, this may help: https://www.jetbrains.com/help/dotmemory/Analyzing_GC_Roots.html.

**Reply by mysteryx93** [reply]
dotMemory is now included in JetBrains Rider... really can't figure out how to use it...

**Reply by maxkatz6** [reply]
dotMemory as a standalone tool was more powerful last time I checked.

**Reply by mysteryx93** [reply]
ya now it's integrated but most of the doc is about a standalone tool that no longer exists
